### ISSUE TITLE
Error building cpymad on M1 Apple Silicon - "-march=core2" not supported

### ISSUE THREAD

**Issue by GuillaumeRD** [hibtc/cpymad#120]
I followed the new installation procedure for Apple Silicon (M1 in this case) and successfully built MAD-X, however when running:

> python setup.py build_ext -lblas -llapack

I get the following error message:

> running build_ext
> building 'cpymad.libmadx' extension
> clang-14 -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /Users/[$username]/miniforge3/envs/xsuite-arm/include -arch arm64 -fPIC -O2 -isystem /Users/[$username]/miniforge3/envs/xsuite-arm/include **-arch arm64 -march=core2 -mtune=haswell** -mssse3 -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -isystem /Users/[$username]/miniforge3/include -D_FORTIFY_SOURCE=2 -isystem /Users/[$username]/miniforge3/include -I/[$MADXDIR]/include -I/Users/[$username]/miniforge3/envs/xsuite-arm/include/python3.10 -c src/cpymad/libmadx.c -o build/temp.macosx-11.0-arm64-cpython-310/src/cpymad/libmadx.o -std=gnu99
> clang-14: error: the clang compiler does not support '-march=core2'
> error: command '/Users/[$username]/miniforge3/envs/xsuite-arm/bin/clang-14' failed with exit code 1

Some context:
* `[$username]` and `[$MADXDIR]` are generic names I am using here instead of the actual paths on the disk;
* I have tried different compilers as CC (see installation instructions): gcc, clang, and clang-14 - but they all return the same error message above;
* I put in bold a part of the error message that I find weird: if the call to clang knows that the architecture of the machine is Silicon as seen with the `-arch arm64`, why is the compilation still calling for the `-march` and `-mtune` flags? Any chance I can edit those out somewhere and replace them both with `-mcpu=apple-m1` as I was able to do when building MAD-X? So far I have not been able to find where those flags are available...

**Comment by coldfix** [comment]
Hey,

> I followed the new installation procedure for Apple Silicon

you mean the one outlined in #114?

I'm not good with Apple, but regarding the last part of the question:

> Any chance I can edit those out somewhere and replace them both with -mcpu=apple-m1 as I was able to do when building MAD-X?

You can try simply invoking the compiler manually, just copy the whole line and replace what you think needs to be changed. If you succeed in building the `.so` file and place in the correct location under `build/` as given by the command line above, `setup.py` will automatically pick up the file without trying to rebuild (if it's newer than the `.c` file and so on). That's also the procedure we're currently using for makig the windows wheels.

I'm currently also working on getting M1 wheels via cross-compilation as recommended in #113.

**Comment by GuillaumeRD** [comment]
I actually followed the instructions listed here:

[https://xsuite.readthedocs.io/en/latest/installation.html#install-on-apple-silicon](url)

along with what can be found here:

[https://github.com/hibtc/cpymad/blob/master/doc/installation/macos.rst](url)

since the mention of running `pip install -U cython wheel setuptools delocate` could not be found at the first link and made one of the scripts look for `libmadx.c`  - which is not available through `git clone`, only `libmadx.pyx` can be found.

I tried your suggestion above and ran the compiler manually: I got 6 warning messages related to `CYTHON_FALLTHROUGH` and it generated a `libmadx.o` file, not `libmadx.so`. I moved it into the temporary `build/` directory regardless, but upon running `setup.py` it still gives me the `-march` error message as it seems to not see the file I just built...

**Comment by GuillaumeRD** [comment]
Thanks for the detailed reply; turns out that it had to do with the environment the `conda` install had set up: when checking on `env`, I noticed that I had to manually update the following variables

> CFLAGS=-march=core2 -mtune=haswell [...]
> CONDA_BACKUP_DEBUG_CFLAGS=-march=core2 -mtune=haswell [...]
> DEBUG_CFLAGS=-march=core2 -mtune=haswell [...]
> CONDA_BACKUP_CFLAGS=-march=core2 -mtune=haswell [...]

by replacing `-march=core2 -mtune=haswell` with `-mcpu=apple-m1`. I re-did the full `cpymad` install from scratch following your updated commands, and as soon as I changed the 4 `CFLAGS` above the installation proceeded flawlessly. Running the test via `python -m pytest test` returned 67 passed, 2 xpassed, 4 warnings.
